Validate pgp signatures

dignifiedquire opened the pull request:
Ref #2028

kinnison commented on src/dist/download.rs:
```diff
@@ -167,7 +198,14 @@ impl<'a> DownloadCfg<'a> {
+            if let Err(err) = self.check_signature(&url_str, &file) {
+                // TODO: this should probably be warn!, but log is not used in this part of the code
+                // at the moment.
+                println!("Invalid signature detected: {}", err);
```
Actually let's have a notification for this since the approach here in the code is to send a notification.

kinnison commented on src/dist/rust-signing-key.asc:
```diff
@@ -0,0 +1,221 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```
Turns out this is at `src/rust-key.gpg.ascii` too, we should probably not include a second copy.  The config object carries a facility to get the key but I don't like it.  For now please drop this file and include the one already present.

kinnison commented on src/dist/signatures.rs:
```diff
@@ -0,0 +1,51 @@
+use crate::errors::*;
+
+// const SIGNING_KEY_BYTES: &[u8] = include_bytes!("rust-signing-key.asc");
+const SIGNING_KEY_BYTES: &[u8] = include_bytes!("../../tests/mock/signing-key.pub.asc");
```
Eventually we'll drop both of these, but for now change the "real" one (commented out) to pull in the `src/rust-key.gpg.ascii` file mentioned in the previous comment.

kinnison commented on src/dist/signatures.rs:
```diff
@@ -0,0 +1,51 @@
+            .collect::<std::result::Result<Vec<_>, _>>()
+            .map_err(squish_internal_err).unwrap()
+    };
+}
```
When we move to configurable keys, we'll need to load these on the fly, so perhaps don't do this lazy-static?

dignifiedquire replied:
We can always have this one be fixed and avoid the cost of loading it multiple times, while the others are pulled in dynamically

kinnison commented on src/dist/signatures.rs:
```diff
@@ -0,0 +1,51 @@
+    ErrorKind::SignatureVerificationInternalError(format!("{}", err)).into()
+}
+
+pub fn verify_signature(content: &str, signature: &str) -> Result<bool> {
```
Could we please change this so that `content` is given as a `&dyn Read` rather than `&str` ?  I'm thinking ahead to verifying signatures on updates to rustup.

dignifiedquire replied:
sure

kinnison reviewed (changes requested):
This looks amazing as a first-cut.  We need to have tests which verify that if signatures exist which are bad, we get appropriate warnings, and then we need to think about how we're going to handle the keyring more effectively than the current pair of commented lines; but thank you and I hope my suggestions are useful.

dignifiedquire commented:
I will need to wait for a more configurable way for keys to put in the real thing unfortunately, as `cfg(test)` is not applied when running the tests in the `tests` directory (ref likely https://github.com/rust-lang/rust/issues/59168)

bors commented:
:umbrella: The latest upstream changes (presumably #2079) made this pull request unmergeable. Please resolve the merge conflicts.

kinnison reviewed (commented):
This is looking pretty good now.  I'm thinking about how I want to provide key configurability.  I will hopefully have an answer for you in the next few days.

kinnison reviewed (commented):
This is starting to look like it's near to ready.

kinnison commented:
I think we need to start tidying this into a series of functional commits so that we can get ready to merge.  This is looking amazing.

dignifiedquire committed:
feat: implement pgp signature validation

Uses the pgp crate to validate signatures on downloaded artifacts when they are available and warns if those are not valid.

Ref #2028

kinnison reviewed (approved):
That test looks :+1:

dignifiedquire committed:
fixup: use correct path separator

kinnison reviewed (approved):
That makes sense, here's hoping it sorts appveyor out :D

dignifiedquire committed:
fixup: fight those paths
